### PR TITLE
Redis cache plugin

### PR DESCRIPTION
**Pull request by mirceaulinic**
### What does this PR do?

Plugin to cache mine, grains etc in Redis.

Redis is very fast when it comes to retrieve the content of a key, but it's very slow when one needs to build hierarchies. In that case, the user requires to build different structure rather than straightfroward k-v store.
As described in the documentation, each action requires one or more adjustments of the keys in order to be able to rebuild more efficiently the tree-like hierarchy.

In time we may add some features to the `flush` function in order to improve the performances. Right now it is going to suffer when flushing a bank with _many_ sub-banks, or sub-tree consisting on many depth levels.
This can be improved later, through a Lua script loaded on the user's server and just called.

### CODE REVIEW

**Comment by mirceaulinic** [comment]
I have tested this and it feels much faster than localfs.

This is what I have in my local instance (one minion only as this is still running on the testing platform).

```yaml
127.0.0.1:6379> keys *
1) "$KEY_minions/edge01.bjm01/mine"
2) "$BANKEYS_minions/edge01.bjm01"
3) "$KEY_minions/edge01.bjm01/data"
4) "$BANK_minions"
```

File: salt/cache/redis_cache.py [hunk 1]
```diff
@@ -0,0 +1,453 @@
+# -*- coding: utf-8 -*-
+'''
+Redis
+=====
+
+Redis plugin for the Salt caching subsystem.
+
+.. versionadded:: Nitrogen
+
+As Redis provides a simple mecanism for very fast key-value store,
```
Comment (cachedout) [inline salt/cache/redis_cache.py]: `mechanism`

File: salt/cache/redis_cache.py [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+in order to privde the necessary features for the Salt
+caching subsystem, the following conventions are used:
+
+- a Redis key consists of the bank name and the cache key separated by ``/``, e.g.:
+``$KEY_minions/alpha/stuff`` where ``minions/alpha`` is the bank name
+and ``stuff`` is the key name.
+- as the caching subsystem is organised as a tree, we need to store the
+caching path and identify the bank and it's offsprings.
```
Comment (cachedout) [inline salt/cache/redis_cache.py]: its

Comment (cachedout) [inline salt/cache/redis_cache.py]: `offspring`

File: salt/cache/redis_cache.py [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+    if redis_value is None:
+        return redis_value
+    return __context__['serial'].loads(redis_value)
+
+
+def flush(bank, key=None):
+    '''
+    Remove the key from the cache bank with all the key content. If no key is specified, remove
+    the entire bank with all keys and sub-banks inside.
+    This function is using the Redis pipelining for best performances.
```
Comment (cachedout) [inline salt/cache/redis_cache.py]: `performance`

**Review by cachedout — CHANGES_REQUESTED** [review]
Just some minor documentation issues. The rest looks great!

**Comment by mirceaulinic** [comment]
Thanks @cachedout - I have corrected the typos.

**Pull request merged**